### PR TITLE
Add list around iterator in handle_missing_dep

### LINKED ISSUE #4371: "Set changed size during iteration" Error in Worker.put_key_in_memory()
I am getting an error in this line: https://github.com/dask/distributed/blob/d54388c92f5cb6643180749f7d30f580609379a2/distributed/worker.py#L1970

Maybe a temporary copy of ts.dependents helps here?

```
Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOMainLoop object at 0x7fa72878fb90>>, <Task finished coro=<Worker.gather_dep() done, defined at [...]/python3.7/site-packages/distributed/worker.py:2000> exception=RuntimeError('Set changed size during iteration')>)
Traceback (most recent call last):
File "[...]/python3.7/site-packages/tornado/ioloop.py", line 743, in _run_callback
  ret = callback()
File "[...]/python3.7/site-packages/tornado/ioloop.py", line 767, in _discard_future_result
  future.result()
File "[...]/python3.7/site-packages/distributed/worker.py", line 2119, in gather_dep
  self.transition(ts, "memory", value=data[d])
File "[...]/python3.7/site-packages/distributed/worker.py", line 1539, in transition
  state = func(ts, **kwargs)
File "[...]/python3.7/site-packages/distributed/worker.py", line 1605, in transition_flight_memory
  self.put_key_in_memory(ts, value)
File "[...]/python3.7/site-packages/distributed/worker.py", line 1970, in put_key_in_memory
  for dep in ts.dependents:
RuntimeError: Set changed size during iteration
```

### PR DESCRIPTION
**Pull request by mrocklin**
Fixes #4371

- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `black distributed` / `flake8 distributed` / `isort distributed`

### CODE REVIEW

**Comment by gerrymanoim** [comment]
I'm actually not sure how this ever worked. Wouldn't this always lead to a `RuntimeError: Set changed size during iteration`

```
            for dep in deps:
                if dep.suspicious_count > 5:
                    deps.remove(dep)
```

**Comment by jrbourbeau** [comment]
> I'm actually not sure how this ever worked. Wouldn't this always lead to a `RuntimeError: Set changed size during iteration`

That's a fair point. Over in https://github.com/jrbourbeau/distributed/commit/f2f39f15007b73b00dab90a58fec5f6436ea54f1 I added a `breakpoint` right before we remove the bad dependency from `deps` and all tests on that commit passed. So I suspect we don't have test coverage for this code path, which is why the `RuntimeError` had gone largely unnoticed. 

As mentioned in https://github.com/dask/distributed/issues/4371#issuecomment-762979029, there may be other underlying issues (xref https://github.com/dask/distributed/issues/4439), but making a copy, as is done in this PR, is a reasonable improvement on the current situation. 

Do the changes here improve your previous workload https://github.com/dask/distributed/issues/4371#issuecomment-907351285?

File: distributed/worker.py [hunk 1]
```diff
@@ -2574,7 +2574,7 @@ async def handle_missing_dep(self, *deps, **kwargs):
             who_has = {k: v for k, v in who_has.items() if v}
             self.update_who_has(who_has)
             still_missing = set()
-            for dep in deps:
+            for dep in list(deps):
```
Comment (jrbourbeau) [inline distributed/worker.py:2577]: I'm wondering what prompted this `list(...)` here. I could be missing something, but I don't see where the `deps` set gets mutated inside this `for`-loop

**Comment by gerrymanoim** [comment]
It is hard to tell for sure as I don't have a good test harness for this (I can just rerun the workflow that triggered this previously), but this appears to have fixed the issue.

**Comment by gerrymanoim** [comment]
> So I suspect we don't have test coverage for this code path

Is there a helpful PR I could make about this? I see there's an old https://github.com/dask/distributed/blob/main/.coveragerc file and I've used coverage.py  (https://coverage.readthedocs.io/en/coverage-5.5/) to generate html and debug things locally, but I'm not sure if this would be helpful here.

File: distributed/worker.py (hunk 1 above)
Comment (jrbourbeau) [reply distributed/worker.py]: I went ahead and removed this, we can always add it back if needed

**Comment by jrbourbeau** [comment]
Thanks for trying this out @gerrymanoim, I appreciate it. I'll plan to merge this PR after CI finishes up 

Re: test coverage, there's this PR https://github.com/dask/distributed/pull/4670 which I just revived. I agree getting automated coverage running would be a nice improvement.

**Review by jrbourbeau — APPROVED** [review]
Thanks @mrocklin, this is in

**Pull request merged**